runtime/platform/test/executor_pal_runtime_override_test.cpp

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <gtest/gtest.h>
+
+#include <executorch/runtime/platform/platform.h>
+#include <executorch/runtime/platform/test/pal_spy.h>
+
+namespace {
+PalSpy* active_spy;
+
+void pal_init(void) {
+  active_spy->init();
+}
+
+et_timestamp_t pal_current_ticks(void) {
+  return active_spy->current_ticks();
+}
+
+et_tick_ratio_t pal_ticks_to_ns_multiplier(void) {
+  return active_spy->ticks_to_ns_multiplier();
+}
+
+void pal_emit_log_message(
+    et_timestamp_t timestamp,
+    et_pal_log_level_t level,
+    const char* filename,
+    ET_UNUSED const char* function,
+    size_t line,
+    const char* message,
+    ET_UNUSED size_t length) {
+  active_spy->emit_log_message(
+      timestamp, level, filename, function, line, message, length);
+}
+
+void* pal_allocate(size_t size) {
+  return active_spy->allocate(size);
+}
+
+void pal_free(void* ptr) {
+  active_spy->free(ptr);
+}
+} // namespace
+
+class RuntimePalOverrideTest : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    // Capture the current PAL implementation so that it can be restored
+    // after the test.
+    _original_pal_impl = *executorch::runtime::get_pal_impl();
+  }
+
+  void TearDown() override {
+    // Restore the original PAL implementation.
+
+    // This is a slightly hacky way to allow this test to exist alongside
+    // the executor_pal_override_test, which provides a build-time override
+    // for et_pal_init. This implementation asserts that an intercept exists.
+    // Since register_pal calls init, we need to make sure that an intercept
+    // is registered. It will be deregistered when it goes out of scope,
+    // allowing the tests to run in any order.
+    InterceptWith iw(_spy);
+    auto success = executorch::runtime::register_pal(_original_pal_impl);
+    if (!success) {
+      throw std::runtime_error("Failed to restore PAL implementation.");
+    }
+  }
+
+  void RegisterSpy() {
+    active_spy = &_spy;
+
+    executorch::runtime::register_pal(executorch::runtime::pal_table(
```

Comment:
Thinking out loud, the PAL resolution earlier would happen at compile/link time, worst case at load time through weak symbol resolution. While with C++ static init (before we enter main) or worst manual call during init would happen way after the process starts running. So if there are things we want to intercept, we can't use stuff from PAL like timestamp or print calls, this is very important for something like Linux kernel which jumps through some hoops to get printk enabled as soon as possible.

I wonder if this is important to us, not now but also in the future looking at bare metal deployments, not so much for rich OSes.

Reply:
Makes sense. Ultimately, the existing weak mechanism still works and will be set up at link time. Tbh what I'd prefer is if ET had an explicit init call where we could pass in library config, but it's probably a bit late for that.